### PR TITLE
Add Panfrost support

### PR DESCRIPTION
**Pull request by adelyser**
Add Panfrost support for the CB1 and BTT PAD7.

Compile with `CONFIG_DRM_PANFROST` enabled.

Instructions for installing the userspace binaries are here:
https://gist.github.com/adelyser/e814dd7737026765a3ab390ec21f78cc

Screenfetch shows the `Mali G-31 Panfrost` GPU
Screenshot from 2023-06-07 19-03-13

`glmark2-es2-wayland` results in benchmark score of 273 when running Gnome.
Screenshot from 2023-06-07 18-53-59
Screenshot from 2023-06-07 18-54-49

### CODE REVIEW

File: kernel/drivers/gpu/drm/Makefile [hunk 1]
```diff
@@ -40,13 +40,20 @@ drm_kms_helper-y := drm_bridge_connector.o drm_crtc_helper.o \
 
 drm_kms_helper-$(CONFIG_DRM_FBDEV_EMULATION) += drm_fb_helper.o
 drm_kms_helper-$(CONFIG_DRM_KMS_CMA_HELPER) += drm_fb_cma_helper.o
+drm-$(CONFIG_DRM_GEM_SHMEM_HELPER) += drm_gem_shmem_helper.o
 
 obj-$(CONFIG_DRM_KMS_HELPER) += drm_kms_helper.o
+obj-$(CONFIG_DRM_VGEM)	+= vgem/
+obj-$(CONFIG_DRM_VKMS)	+= vkms/
+obj-$(CONFIG_DRM_MESON)	+= meson/
```
Comment (bigtreetech) [inline kernel/drivers/gpu/drm/Makefile:48]: There is no `meson` folder

File: kernel/drivers/gpu/drm/Makefile [hunk 2]
```diff
@@ -40,13 +40,20 @@ drm_kms_helper-y := drm_bridge_connector.o drm_crtc_helper.o \
...
 
 obj-$(CONFIG_DRM)	+= drm.o
 
 obj-$(CONFIG_DRM_PANEL_ORIENTATION_QUIRKS) += drm_panel_orientation_quirks.o
+obj-$(CONFIG_DRM_SCHED)	+= scheduler/
 
 obj-$(CONFIG_DRM_SUN4I) += sun4i/
 
 obj-y			+= bridge/
+obj-$(CONFIG_DRM_LIMA)  += lima/
```
Comment (bigtreetech) [inline kernel/drivers/gpu/drm/Makefile:58]: There is no `lima` folder

**Comment by bigtreetech** [comment]
`obj-$(CONFIG_DRM_MESON)	+= meson/`
and
`obj-$(CONFIG_DRM_LIMA)  += lima/`
will cause a compilation error, can it be deleted ?

**Comment by adelyser** [comment]
> `obj-$(CONFIG_DRM_MESON) += meson/` and `obj-$(CONFIG_DRM_LIMA) += lima/` will cause a compilation error, can it be deleted ?

Good spotting, yes, those can be removed since the folders do not exist. There cannot be a compilation error since those definitions are not in the Kconfig, so they aren't selectable, but they still shouldn't be there. Thanks for finding that!

**Comment by bigtreetech** [comment]
Got it.

**Comment by bigtreetech** [comment]
@adelyser 
Hi, Recently, when testing KlipperScreen, it was found that the GPU may cause rotation failure https://github.com/KlipperScreen/KlipperScreen/issues/1035

When set to `left` or `right`, there are the following errors in `dmesg`
```
[  153.316434] panfrost 1800000.gpu: js fault, js=0, status=DATA_INVALID_FAULT, head=0x308c980, tail=0x308c980
[  153.316800] panfrost 1800000.gpu: js fault, js=0, status=IMPRECISE_FAULT, head=0x3135980, tail=0x3135980
[  153.337053] panfrost 1800000.gpu: js fault, js=0, status=IMPRECISE_FAULT, head=0x3135980, tail=0x3135980
[  157.910864] panfrost 1800000.gpu: js fault, js=0, status=DATA_INVALID_FAULT, head=0x308ca00, tail=0x308ca00
[  157.911118] panfrost 1800000.gpu: js fault, js=0, status=IMPRECISE_FAULT, head=0x318c980, tail=0x318c980
[  158.017882] panfrost 1800000.gpu: js fault, js=0, status=DATA_INVALID_FAULT, head=0x308ca00, tail=0x308ca00
[  158.018064] panfrost 1800000.gpu: js fault, js=0, status=IMPRECISE_FAULT, head=0x318c980, tail=0x318c980

```

do you have any suggestions?

**Comment by adelyser** [comment]
> @adelyser Hi, Recently, when testing KlipperScreen, it was found that the GPU may cause rotation failure [KlipperScreen/KlipperScreen#1035](https://github.com/KlipperScreen/KlipperScreen/issues/1035)
> 
> When set to `left` or `right`, there are the following errors in `dmesg`
> 
> ```
> [  153.316434] panfrost 1800000.gpu: js fault, js=0, status=DATA_INVALID_FAULT, head=0x308c980, tail=0x308c980
> [  153.316800] panfrost 1800000.gpu: js fault, js=0, status=IMPRECISE_FAULT, head=0x3135980, tail=0x3135980
> [  153.337053] panfrost 1800000.gpu: js fault, js=0, status=IMPRECISE_FAULT, head=0x3135980, tail=0x3135980
> [  157.910864] panfrost 1800000.gpu: js fault, js=0, status=DATA_INVALID_FAULT, head=0x308ca00, tail=0x308ca00
> [  157.911118] panfrost 1800000.gpu: js fault, js=0, status=IMPRECISE_FAULT, head=0x318c980, tail=0x318c980
> [  158.017882] panfrost 1800000.gpu: js fault, js=0, status=DATA_INVALID_FAULT, head=0x308ca00, tail=0x308ca00
> [  158.018064] panfrost 1800000.gpu: js fault, js=0, status=IMPRECISE_FAULT, head=0x318c980, tail=0x318c980
> ```
> 
> do you have any suggestions?

Those errors are typically caused by not having the proper user space binaries installed. Can you try installing them using the gist link in the PR description? That should resolve the issues.

Bullseye doesn’t have new enough versions of the libraries, you’ll need to either build them or update to bookworm.

**Comment by bigtreetech** [comment]
Got. fixed by the steps in gist link.
Thank you very much!

**Comment by VilriginMCH** [comment]
Hello, I successfully built and installed the KDE and Mesa on my CB1 board with the instruction in gist link, and the dmesg output shows info of the panfrost driver like this:
```
[   11.154492] panfrost 1800000.gpu: clock rate = 432000000
[   11.154549] panfrost 1800000.gpu: bus_clock rate = 200000000
[   11.185082] panfrost 1800000.gpu: mali-g31 id 0x7093 major 0x0 minor 0x0 status 0x0
[   11.185115] panfrost 1800000.gpu: features: 00000000,3fde77ff, issues: 00000000,00000400
[   11.185127] panfrost 1800000.gpu: Features: L2:0x07100206 Shader:0x00000000 Tiler:0x00000209 Mem:0x1 MMU:0x00002821 AS:0xff JS:0x7
[   11.185139] panfrost 1800000.gpu: shader_present=0x1 l2_present=0x1
[   11.190166] [drm] Initialized panfrost 1.2.0 20180908 for 1800000.gpu on minor 1
```
However, the screenfetch command shows that the GPU is  Software Rasterizer, which proves that the GPU doesn't work. I have no idea on it.
How can I figure out and solve this problem? Thanks a lot!

**Pull request merged**